storage/src/main/java/org/apache/kafka/storage/internals/epoch/LeaderEpochFileCache.java

```diff
@@ -0,0 +1,403 @@
+            }
+        } finally {
+            lock.writeLock().unlock();
+        }
+    }
+
+    /**
+     * Clears old epoch entries. This method searches for the oldest epoch < offset, updates the saved epoch offset to
+     * be offset, then clears any previous epoch entries.
+     * <p>
+     * This method is exclusive: so truncateFromStart(6) will retain an entry at offset 6.
+     *
+     * @param startOffset the offset to clear up to
+     */
+    public void truncateFromStart(long startOffset) {
+        lock.writeLock().lock();
+        try {
+            List<EpochEntry> removedEntries = removeFromStart(entry -> entry.startOffset <= startOffset);
+
+            if (!removedEntries.isEmpty()) {
+                EpochEntry firstBeforeStartOffset = removedEntries.get(removedEntries.size() - 1);
+                EpochEntry updatedFirstEntry = new EpochEntry(firstBeforeStartOffset.epoch, startOffset);
+                epochs.put(updatedFirstEntry.epoch, updatedFirstEntry);
+
+                flush();
+
+                log.debug("Cleared entries {} and rewrote first entry {} after truncating to start offset {}, leaving {} in the cache.", removedEntries, updatedFirstEntry, startOffset, epochs.size());
+            }
+        } finally {
+            lock.writeLock().unlock();
+        }
+    }
+
+    public OptionalInt epochForOffset(long offset) {
+        lock.readLock().lock();
+        try {
+            Integer previousEpoch = null;
+            for (EpochEntry epochEntry : epochs.values()) {
+                int epoch = epochEntry.epoch;
+                long startOffset = epochEntry.startOffset;
+
+                // Found the exact offset, return the respective epoch.
+                if (startOffset == offset) return OptionalInt.of(epoch);
+
+                // exit from the loop as it is more than the target offset
+                if (startOffset > offset) break;
+
+                previousEpoch = epoch;
+            }
+
+            return previousEpoch != null ? OptionalInt.of(previousEpoch) : OptionalInt.empty();
+        } finally {
+            lock.readLock().unlock();
+        }
+    }
+
+    /**
+     * Delete all entries.
+     */
+    public void clearAndFlush() {
+        lock.writeLock().lock();
+        try {
+            epochs.clear();
+            flush();
+        } finally {
+            lock.writeLock().unlock();
+        }
+    }
+
+    public void clear() {
+        lock.writeLock().lock();
+        try {
+            epochs.clear();
+        } finally {
+            lock.writeLock().unlock();
+        }
+    }
+
+    // Visible for testing
+    public List<EpochEntry> epochEntries() {
+        lock.writeLock().lock();
+        try {
+            return new ArrayList<>(epochs.values());
+        } finally {
+            lock.writeLock().unlock();
+        }
+    }
+
+    public void flush() {
+        lock.writeLock().lock();
```

Comment:
We added a lock here that wasn't there before. It looks like this is only used within this class. Can we make it private and remove the lock?

Reply:
We should have a lock here as this method is accessing epoch entries which can be added/removed through other methods from different threads.

Reply:
Yes, but isn't `flush` only called with the lock held? Looking at the usage more closely, this is the case for every usage except `assign`. So, it does look like we have a thread safety bug for the `assign` case. However, the change here has a couple of issues:

Reply:
Also, it's not clear why a `write` lock is required for this operation since we are only reading from `epochs` (the actual checkpoint write has a separate lock).

Reply:
Thinking about it, how about we switch the write lock to a read lock and leave the rest as it is?

Reply:
The reason for using write lock is not to allow multiple threads to write concurrently. But we can have a constraint on `LeaderEpochCheckpoint#write` implementation to be thread safe and use a read lock here. So, as you suggested, taking a read lock here should be sufficient with the mentioned constraint.

Reply:
Let's go with that then as it is least likely to cause a change in behavior from what we had in the Scala code.